pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/adminoperation/ParticipantDataFixupService.java

```diff
@@ -0,0 +1,214 @@
+package org.broadinstitute.dsm.service.adminoperation;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.gson.Gson;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.broadinstitute.dsm.db.DDPInstance;
+import org.broadinstitute.dsm.db.dao.ddp.participant.ParticipantDataDao;
+import org.broadinstitute.dsm.db.dto.ddp.participant.ParticipantData;
+import org.broadinstitute.dsm.exception.DSMBadRequestException;
+import org.broadinstitute.dsm.exception.DsmInternalError;
+import org.broadinstitute.dsm.model.defaultvalues.ATDefaultValues;
+import org.broadinstitute.dsm.pubsub.WorkflowStatusUpdate;
+import org.broadinstitute.dsm.service.admin.AdminOperation;
+import org.broadinstitute.dsm.service.admin.AdminOperationRecord;
+
+/**
+ * AdminOperation service to asynchronously fix various ParticipantData issues
+ */
+@Slf4j
+public class ParticipantDataFixupService implements AdminOperation {
+
+    protected enum UpdateStatus {
+        UPDATED,
+        NOT_UPDATED,
+        ERROR,
+        NO_PARTICIPANT_DATA
+    }
+
+    private static final Gson gson = new Gson();
+    protected List<String> validRealms = List.of("atcp");
```

Comment:
Just a thought.
If the service is specific to only ATCP, maybe the name should reflect it like ATParticipantDataFixupService.java
OR have ATParticipantDataFixupService.java as separate class extending ParticipantDataFixupService.java supporting only ATCP and have realm check in there.

Reply:
Yup, at this point it's hard to say what the next ParticipantData fixup would look like, and we may have no other cases, or we could have like 5.

What I tried to do here is strike a balance: The service is parameterized (at the HTTP request level) so we can use that to indicate the type of fixup, but of course there is only one now. I think if the next fixup is a lot like this one (uses much of the same code) then I would probably just keep the structure as is and put conditionals in the code. If the next case is different enough to warrant a separate class, then likely we would implement a class hierarchy as you describe, or perhaps just go with two separate (but renamed) classes.

I am working under the "you're not gonna need it (YAGNI)" principal, and I don't think this implementation has ruled out any future options.

Does that make sense? What do you think?

Reply:
sounds ok to me.